backtesting/backtesting.py

```diff
@@ -1221,101 +1236,172 @@ def maximize(stats: pd.Series, _key=maximize):
+                # No trade was made in any of the runs. Just make a random
+                # run so we get some, if empty, results
+                self.run(**param_combos[0])  # type: ignore
+            else:
+                # Re-run best strategy so that the next .plot() call will render it
+                self.run(**dict(zip(heatmap.index.names, best_params)))
+
+            if return_heatmap:
+                return self._results, heatmap
+            return self._results
+
+        def _optimize_skopt() -> Union[pd.Series, Tuple[pd.Series, pd.Series]]:
+            """
+            Optimize strategy parameters to an optimal combination using
+            scikit-optimize. Returns result `pd.Series` of
+            the best run.
+
+            `maximize` is a string key from the
+            `backtesting.backtesting.Backtest.run`-returned results series,
+            or a function that accepts this series object and returns a number;
+            the higher the better. By default, the method maximizes
+            Van Tharp's [System Quality Number](https://google.com/search?q=System+Quality+Number).
+
+            Additional keyword arguments represent strategy arguments with
+            list-like collections of possible values. For example, the following
+            code finds and returns the "best" of the 7 admissible (of the
+            9 possible) parameter combinations:
+
+                backtest.optimize(sma1=[5, 10, 15], sma2=[10, 20, 40],
+                                constraint=lambda p: p.sma1 < p.sma2)
+
+            .. TODO::
+                Improve multiprocessing with joblib.parallel .
+            """
+            dimensions = []
+            for key, value in kwargs.items():
+                args_array = np.array(value)
+
+                if args_array.dtype.kind == 'i':
+                    dimensions.append(Integer(name=key, low=min(value), high=max(value)))
+                elif args_array.dtype.kind == 'f':
+                    dimensions.append(Real(name=key, low=min(value), high=max(value)))
+                else:
+                    dimensions.append(Categorical(args_array.tolist(), name=key))
+
+            # @ use_named_args(dimensions=dimensions)
+            # def convert(n1, n2):
+            #     res = self.run(**dict(zip(['n1', 'n2'], [n1, n2])))
+            #     return -res[maximize_key]
+            @ use_named_args(dimensions=dimensions)
+            def convert(**params):
+                index = []
+                for item in params:
+                    index.append("'" + item + "'")
+                res = self.run(**dict(zip(index, params)))
+                return -res[maximize_key]
+
+            res = forest_minimize(func=convert,
+                                  dimensions=dimensions,
+                                  n_calls=max_tries, base_estimator="ET",
+                                  random_state=4)
+
+            self.run(**dict(zip(['n1', 'n2'], res.x)))
```

Comment:
This, working with `n1` and `n2`, doesn't produce the right `self._results` in the general case?

Reply:
No it doesn't. I am assuming there could be variable number of features a user might want to optimize the strategy with, so something like **params would be more suitable?

Reply:
Ideally, best evaluated **params:
https://github.com/kernc/backtesting.py/blob/191f4753b2a405b20e30add2267afc69a5b06858/backtesting/backtesting.py#L1311-L1312